src/mixings/logit_sb_mixing.h

```diff
@@ -0,0 +1,77 @@
+#ifndef BAYESMIX_MIXINGS_LOGIT_SB_MIXING_H_
+#define BAYESMIX_MIXINGS_LOGIT_SB_MIXING_H_
+
+#include <google/protobuf/message.h>
+
+#include <Eigen/Dense>
+#include <memory>
+
+#include "base_mixing.h"
+#include "conditional_mixing.h"
+#include "mixing_id.pb.h"
+#include "mixing_prior.pb.h"
+#include "src/hierarchies/base_hierarchy.h"
+
+class LogitSBMixing : public ConditionalMixing {
+ public:
+  struct State {
+    Eigen::MatrixXd regression_coeffs;
+  };
+
+ protected:
+  unsigned int num_clusters;
+  unsigned int dim;
+  State state;
+  Eigen::MatrixXd precision;  // TODO should it be in the state?
+
+  //!
+  void create_empty_prior() override { prior.reset(new bayesmix::LogSBPrior); }
+  //!
+  std::shared_ptr<bayesmix::LogSBPrior> cast_prior() {
+    return std::dynamic_pointer_cast<bayesmix::LogSBPrior>(prior);
+  }
+  //!
+  void initialize_state() override;
+  //!
+  double sigmoid(const double x) const { return 1.0 / (1.0 + std::exp(-x)); }
+  //!
+  Eigen::VectorXd grad_log_full_cond(const Eigen::VectorXd &alpha,
+                                     const std::vector<bool> &is_curr_clus,
+                                     const std::vector<bool> &is_prev_clus);
+  //!
+  double log_like(const Eigen::VectorXd &alpha,
+                  const std::vector<bool> &is_curr_clus,
+                  const std::vector<bool> &is_prev_clus);
+
+ public:
+  // DESTRUCTOR AND CONSTRUCTORS
+  ~LogitSBMixing() = default;
+  LogitSBMixing() = default;
+
+  Eigen::VectorXd get_weights(const Eigen::VectorXd &covariate) const override;
+
+  std::shared_ptr<BaseMixing> clone() const override {
+    return std::make_shared<LogitSBMixing>(*this);
+  }
+
+  //! Returns true if the hierarchy has covariates i.e. is a dependent model
+  bool is_dependent() const override { return true; }
+
+  //!
+  void initialize(const unsigned int n_clust = 1) override;
```

Comment:
I’m not happy with this design choice.. you’re basically saying that nclust here has the same role of nclust in the Dirichlet Mixing, while in the latter case that is truly only used for initialization, here nclust is a crucial parameter. I think it deserves a separate setter / to be in the constructor. the user must be explicit in setting this parameter, it should not be done by the Algorithm

Reply:
Even better, this should be part of the Prior proto!

Reply:
It is indeed set by the user. At line 72 of `base_algorithm.cc` we have `mixing->initialize(init_num_clusters);`. Remember that `init_num_clusters` is one of arguments of `run.cc`.

Reply:
I see your point, but still `init_num_cluster` is not the right parameter name.
For the DPM, PYM that is entirely fine, it's up to the algorithm to select the initial number of cluster, then the posterior will converge to its stationary distribution and that's fine.
For LSB and truncated stick-breaking prior, the `n_clust` (which should be better called `n_components` or `n_atoms`) is part of the model specification. For me, the fact that the algorithm can change this parameter, it would be just like if the algorithm could change the base measure/prior in a mixture model, or even change the hierarchy type, during initialization.

Reply:
`init_num_clusters` is given by the user. The algorithm just assigns it to the mixture through its initialize method, and has no control over it in the current version of the code, nor does it select anything (except interpreting the zero -- an impossible value -- as "one datum per cluster"). It's really no different than using any other setter. If this is an out-of-principle argument, I can create a setter that does the exact same thing.

You may argue `init_num_clusters` is an inappropriate name in this case, and I would agree. It was the correct name for previously-implemented marginal algorithms.

But if we were to put it into a prior object, we would have the absurdity of two potentially different input parameters that refer to the exact same propoerty and which must be the same at all times without any way to enforce this. If what you're suggesting is removing `init_num_clusters` altogether from the main arguments, it can be done, but it may be a lot of work.

Reply:
My point is about modularity, the `num_clusters` parameter is defining the model as much as the choice of hierarchy, is not any less important than that. Yet, it is absurd to think that the choice of the hierarchy passes through the algorithm...
So I would say to forget about the main at the moment, we're building a library, not an executable. We could also have multiple main files if that makes life easier... 

Right now I just want to get the modularity structure right, so that we can replicate that on the Python side!!

Reply:
> the `num_clusters` parameter is defining the model as much as the choice of hierarchy, is not any less important than that

You're right, so much so that we have a unique setter `set_initial_clusters()` that sets both the hierarchy type and the number of clusters at the same time. If the algorithm setting the hierarchy does not defy modularity, neither does the algorithm setting the number of clusters. It just places to the appropriate space what it was given by the user.

If you're still not happy with the current way of setting this value, feel free to suggest one.

Reply:
You're missing my point: right now we have two phases:

1- create the objects that define the statistical model (likelihood + prior)
2- pass these objects to the algorithm in charge of doing MCMC

Now, while in the DPM/PYM the `n_clust` parameter belongs to the algorithm, in the LSB this parameter belongs to the statistical model. Moreover, an `LSB` instance without a num_clusters / num_components parameter should not be allowed.
Hence, since we're using protos to define the statistical model, in my opinion it makes sense to move this parameter to the proto defining the LSB prior.

If you're problem is the compatibility with the `BaseAlgorithm::set_initial_clusters`, it is sufficient to override it in the `BlockedGibbAlgorithm` and print out a warning that the parameter will be ignored.

Reply:
Ok, fair enough. I'll get to work.

Reply:
I think the cleanest solution is to have a `AlgorithmParams` proto message, which collects all the possible params, no need for fancy oneofs here, just one large list of possible parameters would suffice in my opinion.

Reply:
Deal! Maybe I will work on this in a separate future PR.